tests/firedrake/submesh/test_submesh_interpolate.py

```diff
@@ -268,3 +272,75 @@ def expr(m):
     dg2d_ = Function(DG2d).interpolate(dg3d)
     error = assemble(inner(dg2d_ - expr(subm), dg2d_ - expr(subm)) * dx)**0.5
     assert abs(error) < 1.e-14
+
+
+@pytest.mark.parallel(nprocs=[1, 3])
+@pytest.mark.parametrize('fe_fesub', [[("DG", 2), ("DG", 1)],
+                                      [("CG", 3), ("CG", 2)]])
+def test_submesh_interpolate_adjoint(fe_fesub):
+    (family, degree), (family_sub, degree_sub) = fe_fesub
+
+    mesh = UnitSquareMesh(8, 8)
+    x, y = SpatialCoordinate(mesh)
+    subdomain_cond = conditional(LT(x, 0.5), 1, 0)
+    label_value = 999
+    subm = make_submesh(mesh, subdomain_cond, label_value)
+
+    V1 = FunctionSpace(subm, family_sub, degree_sub)
+    V2 = FunctionSpace(mesh, family, degree)
+
+    x, y = SpatialCoordinate(V1.mesh())
+    expr = x * y
+    u1 = Function(V1).interpolate(expr)
+    ustar2 = assemble(inner(1, TestFunction(V2))*dx(label_value))
+
+    expected = assemble(inner(1, u1)*dx(label_value))
+
+    # Test forward 2-form
+    I = assemble(interpolate(TrialFunction(V1), TestFunction(V2.dual()), allow_missing_dofs=True))
+    assert I.arguments()[0].function_space() == V2.dual()
+    assert I.arguments()[1].function_space() == V1
+
+    result_forward_2 = assemble(action(ustar2, action(I, u1)))
+    assert np.isclose(result_forward_2, expected)
+
+    # Test adjoint 2-form
+    I_adj = assemble(interpolate(TestFunction(V1), TrialFunction(V2.dual()), allow_missing_dofs=True))
+    assert I_adj.arguments()[0].function_space() == V1
+    assert I_adj.arguments()[1].function_space() == V2.dual()
+
+    result_adjoint_2 = assemble(action(action(I_adj, ustar2), u1))
+    assert np.isclose(result_adjoint_2, expected)
+
+    # Test forward 1-form (only in serial for now)
+    if V1.comm.size == 1:
+        # Matfree forward interpolation with Submesh currently fails in parallel.
+        # The ghost nodes of the parent mesh may be redistributed
+        # into different processes as non-ghost dofs of the submesh.
+        # The submesh kernel will write into ghost nodes of the parent mesh,
+        # but this will be ignored in the halo exchange if access=op2.WRITE.
+
+        # See https://github.com/firedrakeproject/firedrake/issues/4483
+
+        Iu1 = assemble(interpolate(u1, TestFunction(V2.dual()), allow_missing_dofs=True))
+        assert Iu1.function_space() == V2
+
+        expected_primal = assemble(action(I, u1))
+        assert np.allclose(Iu1.dat.data, expected_primal.dat.data)
+
+        result_forward_1 = assemble(action(ustar2, Iu1))
+        assert np.isclose(result_forward_1, expected)
```

Comment:
I think I would prefer to have
```py
is_close = np.isclose(result_forward_1, expected)
if comm.size == 1:
    assert is_close
else:
    assert not is_close
```
That way if we fix the bug in the future we will know to come back and fix this test.

Reply:
Should we not then striclty XFAIL the test `if comm.size > 1`?

Reply:
But if we do that then we stop testing other parts of this test that do in fact work. If you wanted to xfail these then they'd need to be broken out of the rest of the test.

Reply:
Done